### PR TITLE
feat: add breeding

### PR DESCRIPTION
**Pull request by PetePetelius**

### CODE REVIEW

File: src/misc/breeding.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+function hasFoodSource(entity: LivingEntity): boolean {
+  const blocksNearEntity = blocksInRadiusOfEntity(entity, 2);
+
+  for (const block of blocksNearEntity) {
+    if (entity instanceof Chicken) {
+      if (block.type === SEED_BOWL) {
+        block.setType(EMPTY_SEED_BOWL);
+        const data = block.blockData;
+        if (data instanceof Waterlogged) data.setWaterlogged(false);
+        block.setBlockData(data);
```
Comment (Laetta) [inline src/misc/breeding.ts:29]: replace setters and getters ->

block.setType(...)   ->  block.type = ...

block.setBlockData(...)  --> block.blockdata = ...

File: src/misc/breeding.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+];
+
+const BREED_WATER_LEVEL = 1; // Max 3
+const BREED_COMPOST_LEVEL = 4; // Max 8
+const BOWL_SEED_AMOUNT = 4;
+
+const SEED_BOWL = Material.DEAD_BRAIN_CORAL_FAN;
+const EMPTY_SEED_BOWL = Material.DEAD_TUBE_CORAL_FAN;
+
+function hasFoodSource(entity: LivingEntity): boolean {
```
Comment (bensku) [inline src/misc/breeding.ts:22]: This consumes the nearby food source, and the name should indicate that.

File: src/misc/breeding.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+          return true;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+function hasWaterSource(entity: LivingEntity): boolean {
```
Comment (bensku) [inline src/misc/breeding.ts:57]: This also consumes water, name should not begin with 'has'.

File: src/misc/breeding.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        block.blockData = cauldronData;
+        playDrinkingEffects(block);
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+function blocksInRadiusOfEntity(entity: LivingEntity, radius: number): Block[] {
```
Comment (bensku) [inline src/misc/breeding.ts:73]: Note to self: unified API for iterating blocks around entity/location.

File: src/misc/breeding.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+registerEvent(EntityBreedEvent, (event) => {
+  if (!ANIMALS_WITH_CUSTOM_BREEDING.includes(event.entityType)) return;
+
+  const baby = event.entity as LivingEntity;
+  const mother = event.mother as LivingEntity;
+
+  const hasFood = hasFoodSource(mother);
+  const hasWater = hasWaterSource(mother);
+
+  if (!hasFood || !hasWater) baby.remove();
```
Comment (bensku) [inline src/misc/breeding.ts:133]: If only food or only water is available, it will be consumed when breeding fails. Is this intentional?

**Review by bensku — CHANGES_REQUESTED** [review]

**Review by bensku — APPROVED** [review]

File: src/misc/breeding.ts [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+registerEvent(EntityBreedEvent, (event) => {
+  if (!BREEDABLE_ANIMALS.has(event.entityType.toString())) return;
+
+  const baby = event.entity as LivingEntity;
+  const mother = event.mother as LivingEntity;
+
+  const isFed = consumeFood(mother);
+  const isHydrated = consumeWater(mother);
+
+  if (!isFed || !isHydrated) baby.remove();
```
[the remark this answers is not in the document]
Comment (Laetta) [reply src/misc/breeding.ts:212]: This was designed to be intentional

**Pull request merged**